sd_webui_bayesian_merger/merger.py

```diff
@@ -41,6 +41,42 @@
     "weighted_double_difference": 5,
 }
 
+NAI_KEYS = {
+    "cond_stage_model.transformer.embeddings.": "cond_stage_model.transformer.text_model.embeddings.",
+    "cond_stage_model.transformer.encoder.": "cond_stage_model.transformer.text_model.encoder.",
+    "cond_stage_model.transformer.final_layer_norm.": "cond_stage_model.transformer.text_model.final_layer_norm.",
+}
+
+
+def fix_clip(model: Dict) -> Dict:
+    if KEY_POSITION_IDS in model:
+        model[KEY_POSITION_IDS] = torch.tensor(
+            [list(range(MAX_TOKENS))], dtype=torch.int64
```

Comment:
This seems more of a bandaid to the problem rather than a solution. Below is the CLIP Position ID and keys from 2 merge s involving 4+ models, without skip position id set to 2. It seems as if something is getting mixed up when the key is being merged. All of the models originally had proper clip positions and values as well.
```
Type: torch.float16
0: 0.00000  0.00000
1: 0.58301  -0.41699
2: 1.16602  -0.83398
3: 1.74902  -1.25098
4: 2.33203  -1.66797
5: 2.91406  -2.08594
6: 3.49805  -2.50195
7: 4.08203  -2.91797
8: 4.66406  -3.33594
9: 5.24609  -3.75391
10: 5.82812  -4.17188
11: 6.41406  -4.58594
12: 6.99609  -5.00391
13: 7.57812  -5.42188
14: 8.16406  -5.83594
15: 8.74219  -6.25781
16: 9.32812  -6.67188
17: 9.91406  -7.08594
18: 10.49219  -7.50781
19: 11.07812  -7.92188
20: 11.65625  -8.34375
21: 12.24219  -8.75781
22: 12.82812  -9.17188
23: 13.40625  -9.59375
24: 13.99219  -10.00781
25: 14.57812  -10.42188
26: 15.15625  -10.84375
27: 15.74219  -11.25781
28: 16.32812  -11.67188
29: 16.90625  -12.09375
30: 17.48438  -12.51562
31: 18.07812  -12.92188
32: 18.65625  -13.34375
33: 19.23438  -13.76562
34: 19.82812  -14.17188
35: 20.40625  -14.59375
36: 20.98438  -15.01562
37: 21.57812  -15.42188
38: 22.15625  -15.84375
39: 22.73438  -16.26562
40: 23.31250  -16.68750
41: 23.93750  -17.06250
42: 24.48438  -17.51562
43: 25.06250  -17.93750
44: 25.65625  -18.34375
45: 26.23438  -18.76562
46: 26.81250  -19.18750
47: 27.40625  -19.59375
48: 27.98438  -20.01562
49: 28.56250  -20.43750
50: 29.15625  -20.84375
51: 29.73438  -21.26562
52: 30.31250  -21.68750
53: 30.90625  -22.09375
54: 31.48438  -22.51562
55: 32.06250  -22.93750
56: 32.65625  -23.34375
57: 33.21875  -23.78125
58: 33.81250  -24.18750
59: 34.40625  -24.59375
60: 34.96875  -25.03125
61: 35.56250  -25.43750
62: 36.15625  -25.84375
63: 36.71875  -26.28125
64: 37.31250  -26.68750
65: 37.90625  -27.09375
66: 38.46875  -27.53125
67: 39.06250  -27.93750
68: 39.65625  -28.34375
69: 40.21875  -28.78125
70: 40.81250  -29.18750
71: 41.40625  -29.59375
72: 41.96875  -30.03125
73: 42.56250  -30.43750
74: 43.15625  -30.84375
75: 43.71875  -31.28125
76: 44.31250  -31.68750
```

```
Type: torch.float16
0: 0.00000  0.00000
1: 0.27979  -0.72021
2: 0.55957  -1.44043
3: 0.83887  -2.16016
4: 1.11914  -2.88086
5: 1.39941  -3.60156
6: 1.67773  -4.32031
7: 2.09375  -4.90625
8: 2.23828  -5.76172
9: 2.51758  -6.48438
10: 2.79883  -7.20312
11: 3.07422  -7.92578
12: 3.35547  -8.64062
13: 3.63672  -9.35938
14: 4.05078  -9.95312
15: 4.19531  -10.80469
16: 4.47656  -11.52344
17: 4.75781  -12.24219
18: 5.03516  -12.96875
19: 5.44922  -13.54688
20: 5.59766  -14.40625
21: 5.87500  -15.12500
22: 6.14844  -15.85156
23: 6.43750  -16.56250
24: 6.71094  -17.28125
25: 6.98828  -18.01562
26: 7.27344  -18.71875
27: 7.55078  -19.45312
28: 7.96484  -20.03125
29: 8.10938  -20.89062
30: 8.39062  -21.60938
31: 8.67188  -22.32812
32: 8.95312  -23.04688
33: 9.35938  -23.64062
34: 9.51562  -24.48438
35: 9.78906  -25.21875
36: 10.07031  -25.93750
37: 10.34375  -26.65625
38: 10.76562  -27.23438
39: 10.90625  -28.09375
40: 11.19531  -28.81250
41: 11.60156  -29.40625
42: 11.75000  -30.25000
43: 12.17188  -30.82812
44: 12.29688  -31.70312
45: 12.58594  -32.40625
46: 12.87500  -33.12500
47: 13.14062  -33.87500
48: 13.42188  -34.56250
49: 13.70312  -35.31250
50: 13.97656  -36.03125
51: 14.26562  -36.75000
52: 14.54688  -37.43750
53: 14.81250  -38.18750
54: 15.10156  -38.90625
55: 15.38281  -39.62500
56: 15.78906  -40.21875
57: 15.93750  -41.06250
58: 16.21875  -41.78125
59: 16.50000  -42.50000
60: 16.78125  -43.21875
61: 17.18750  -43.81250
62: 17.34375  -44.65625
63: 17.62500  -45.37500
64: 17.90625  -46.09375
65: 18.18750  -46.81250
66: 18.59375  -47.40625
67: 18.73438  -48.25000
68: 19.03125  -48.96875
69: 19.28125  -49.71875
70: 19.57812  -50.43750
71: 20.00000  -51.00000
72: 20.14062  -51.87500
73: 20.40625  -52.59375
74: 20.68750  -53.31250
75: 20.98438  -54.00000
76: 21.39062  -54.62500
```
My main worry is that if this is happening to this key, is it possible that similar weirdness is happening to other keys.

Reply:
Actually, I was able to narrow down the issue to the fact that we are weighting the result from the subtracted model on the KEY_position_id key. When you do this, the values go to zero, but we aren't adding up to the proper value because for this key since we are weighting t0 expecting that we are going to get a value from the result of the difference, when in fact we are getting zeros. Here's a comparison from pdb where I weighted t0 with 1-alpha, and one where I didn't for this position id key.
![image](https://user-images.githubusercontent.com/77745122/231231515-eff3a38b-1df4-43b6-bf93-c84b2ec5e6dd.png)
I think the solution is to not weigh t0 for the subtracted models for the position id key, since we have nothing to weigh for since the output of the subtraction will be all zeros anyways.

Reply:
A weighted Approach will prevent that, however you can't apply the weighting in cases when the result of B-C is zero, since we should just take the value of A and override the weight since nothing remained of the tensor after the result of B-C, so a zero check on the tensor created by B-C before merging it should stop this from happening.

Reply:
So a solution could be to put an if statement that uses torch.count_nonzero(input, dim=None) (or the tf equivalent) to check if there are no non-zero values and if so to only return the value of A. However in cases where we are dealing with significantly more complex merges, then we may have to do this on a case by case basis. So for Weighted Diff, you would just return A, for Weighted Double diff, you need to check both B-C and D-E and if one of them contains a non zero tensor while the other is zero then you change the equation to be a normal Weighted Diff and drop the zero tensor, and only weight the other with alpha instead of with beta. Basically like this:

(1-alpha) * t0 + alpha * ( t1-t2 ) : If torch.count_nonzero(t1-t2) == 0 return t0

(1-alpha) * t0 + alpha * ( (1-beta) * (t1 - t2) + beta * (t3-t4) ) : If torch.count_nonzero(t1-t2) == 0 AND torch.count_nonzero(t3-t4) != 0 : return (1-alpha) * t0 + alpha * (t3-t4)

(1-alpha) * t0 + alpha * ( (1-beta) * (t1 - t2) + beta * (t3-t4) ) : If torch.count_nonzero(t1-t2) != 0 AND torch.count_nonzero(t3-t4) == 0 : return (1-alpha) * t0 + alpha * (t1-t2)

Reply:
Welp I was partially wrong, even when dealing with zero tensors it doesn't seem to fix it.

Reply:
> Thing is you may have a zero tensor in the `A` side when alpha is small/large

The way I dealt with it is by just returning A. Also, it's possible that because of the level of precision we have in the values of these tensors, perhaps rounding them to three digits to see if they zero out. I actually never got a zero tensor in my test which was very strange. Perhaps the value just becomes so small that when rounded to three places it zeros out? Because one of the main differences in Supermerger is that when they MBW merge, they round all tensors to three decimal places.
Here's an image from a couple of tensors when subtracted and then rounded to 3 decimal places.
![image](https://user-images.githubusercontent.com/77745122/231827646-43a2a436-4433-4602-99a2-4ca8a43f572d.png)
![image](https://user-images.githubusercontent.com/77745122/231827892-45f552c7-5e44-4635-92e6-d5407cbda0c9.png)

And when doing this on a large scale, there are 179 tensors which record as being zero when rounded to 3 places.
The keys are in this pastebin as I don't want to crowd this comment with them. 
https://pastebin.com/u7cPm9j0
This introduces the problem of, what do we do when we only have 1 or 2 or maybe 3 small values in the difference of t1-t2, how do we properly weight and add those without losing granularity in the other tensor values.